configstack/stack_plan.go

```diff
@@ -0,0 +1,135 @@
+func getResultHandler(detailedExitCode bool, results *[]moduleResult) ModuleHandler {
+	return func(module TerraformModule, output string, err error) error {
+		inspectPlanResult(module, output)
+		if exitCode, convErr := shell.GetExitCode(err); convErr == nil && detailedExitCode && exitCode == 2 {
+			err = nil
+		}
+
+		message, count := extractSummaryResultFromPlan(output)
+
+		// We add the result to the result list (there is no concurrency problem because it is handled by the running_module)
+		*results = append(*results, moduleResult{module, err, message, count})
+
+		return err
+	}
+}
+
+// Print a little summary of the plan execution
+func printSummary(terragruntOptions *options.TerragruntOptions, results []moduleResult) {
+	fmt.Fprintf(terragruntOptions.Writer, "%s\nSummary:\n", separator)
+	for _, result := range results {
+		errMsg := ""
+		if result.Err != nil {
+			errMsg = fmt.Sprintf(", Error: %v", result.Err)
+		}
+
+		fmt.Fprintf(terragruntOptions.Writer, "    %v : %v%v\n", result.Module.Path, result.Message, errMsg)
+	}
+}
+
+// Check the output message
+func inspectPlanResult(module TerraformModule, output string) {
+	if strings.Contains(output, ": Resource 'data.terraform_remote_state.") {
+		var dependenciesMsg string
+		if len(module.Dependencies) > 0 {
+			dependenciesMsg = fmt.Sprintf(" contains dependencies to %v and", module.Config.Dependencies.Paths)
+		}
+		module.TerragruntOptions.Logger.Printf("%v%v refers to remote state "+
+			"you may have to apply your changes in the dependencies prior running terragrunt plan-all.\n",
+			module.Path,
+			dependenciesMsg,
+		)
+	}
+}
+
+// Parse the output message to extract a summary
+func extractSummaryResultFromPlan(output string) (string, int) {
+	const noChange = "No changes. Infrastructure is up-to-date."
+	if strings.Contains(output, noChange) {
+		return "No change", 0
+	}
+
+	re := regexp.MustCompile(`(\d+) to add, (\d+) to change, (\d+) to destroy.`)
+	result := re.FindStringSubmatch(output)
+	if len(result) == 0 {
+		return "Unable to determine the plan status", -1
+	}
+
+	// Count the total number of changes
+	sum := 0
+	for _, value := range result[1:] {
+		count, _ := strconv.Atoi(value)
+		sum += count
+	}
+	if sum != 0 {
+		return result[0], sum
+	}
+
+	// Sometimes, terraform returns 0 add, 0 change and 0 destroy. We return a more explicit message
+	return "No effective change", 0
+}
+
+// This is used to return the total number of changes if -detail-exitcode is specified and return an exit code of 2
+// to be compliant with the terraform specification.
+type countError struct{ count int }
+
+func (err countError) Error() string {
+	article, s := "is", ""
+	if err.count > 1 {
+		article, s = "are", "s"
+	}
+	return fmt.Sprintf("There %s %v change%s to apply", article, err.count, s)
+}
+
+// If there are changes, the exit status must be = 2
+func (err countError) ExitStatus() (int, error) {
+	if err.count > 0 {
+		return 2, nil
```

Comment:
The exit code "magic numbers" (0, 1, 2) are copy/pasted all over the place. Define named constants for them to make the code easier to follow.